azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -0,0 +1,172 @@
+
+        if not self.is_policy_enforcement_supported():
+            msg = "Attempted to enable policy enforcement, but feature is not supported on this platform."
+            self.log_policy(msg=msg)
+            # TODO: surface as a user error with clear instructions for fixing
+            raise PolicyError(msg)
+
+        try:
+            self._engine = regorus.Engine(policy_file=policy_file, rule_file=rule_file)
+            self._policy_engine_enabled = True
+        except Exception as ex:
+            msg = "Failed to initialize Regorus policy engine: '{0}'".format(ex)
+            self.log_policy(msg=msg)
+            raise PolicyError(msg)
+
+    @classmethod
+    def log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_event=True):
+        """
+        Log information to console and telemetry.
+        """
+        if is_success:
+            logger.info(msg)
+        else:
+            logger.error(msg)
+        if send_event:
+            add_event(op=op, message=msg, is_success=is_success)
+
+    @staticmethod
+    def is_policy_enforcement_enabled():
+        """Check whether user has opted into policy enforcement feature"""
+        # TODO: The conf flag will be removed post private preview. Before public preview, add checks
+        # according to the planned user experience (TBD).
+        return conf.get_extension_policy_enabled()
+
+    @staticmethod
+    def is_policy_enforcement_supported():
+        """Check that both platform architecture and distro/version are supported."""
+        osutil = get_osutil()
+        arch = osutil.get_vm_arch()
+        if arch not in POLICY_SUPPORTED_ARCHITECTURE:
+            return False
+        __distro__ = get_distro()
+        DISTRO_NAME = __distro__[0]
+        DISTRO_VERSION = __distro__[1]
+        try:
+            distro_version = DistroVersion(DISTRO_VERSION)
+        except ValueError:
+            raise ValueError
+
+        # Check if the distro is in the support matrix and if the version is supported
+        if DISTRO_NAME in POLICY_SUPPORTED_DISTROS_MIN_VERSIONS:
+            min_version = POLICY_SUPPORTED_DISTROS_MIN_VERSIONS[DISTRO_NAME]
+            return distro_version >= min_version
+        else:
+            return False
+
+    @property
+    def policy_engine_enabled(self):
+        """This property tracks whether the feature is enabled and Regorus engine has been successfully initialized"""
+        return self._policy_engine_enabled
+
+    def evaluate_query(self, input_dict, query):
+        """
+        Input_dict should be a dict. Expected format for input_dict:
+        {
+            "extensions": {
+                "<extension_name_1>": {
+                    "signingInfo": {
+                        "extensionSigned": <true, false>
+                    }
+                }, ...
+        }
+
+        Expected format for query: "data.agent_extension_policy.extensions_to_download"
```

Comment:
what does "data.agent_extension_policy.extensions_to_download"? an extension name?

Reply:
This is the value we're querying for in the regorus engine. We can query for "data", "data.agent_extension_policy", "data.agent_extension_policy.extensions_to_download", etc.

"data.agent_extension_policy.extensions_to_download" will return output in the format:   
```
"extensions_to_download": { 
        "Microsoft.Azure.ActiveDirectory.AADSSHLoginForLinux": { 
            "downloadAllowed": true 
        }, 
        "Microsoft.Azure.Extensions.CustomScript": { 
            "downloadAllowed": false 
        } 
    }
```
...
I've put a full example in the spec.